ssl/quic/quic_fc.c

```diff
@@ -0,0 +1,366 @@
+{
+    int ok = 1;
+    uint64_t credit = ossl_quic_txfc_get_credit_local(txfc);
+
+    if (num_bytes > credit) {
+        ok = 0;
+        num_bytes = credit;
+    }
+
+    if (num_bytes > 0 && num_bytes == credit)
+        txfc->has_become_blocked = 1;
+
+    txfc->swm += num_bytes;
+    return ok;
+}
+
+int ossl_quic_txfc_consume_credit(QUIC_TXFC *txfc, uint64_t num_bytes)
+{
+    int ok = ossl_quic_txfc_consume_credit_local(txfc, num_bytes);
+
+    if (txfc->parent != NULL) {
+        assert(txfc->parent->parent == NULL);
+        if (!ossl_quic_txfc_consume_credit_local(txfc->parent, num_bytes))
+            return 0;
+    }
+
+    return ok;
+}
+
+int ossl_quic_txfc_has_become_blocked(QUIC_TXFC *txfc, int clear)
+{
+    int r = txfc->has_become_blocked;
+
+    if (clear)
+        txfc->has_become_blocked = 0;
+
+    return r;
+}
+
+uint64_t ossl_quic_txfc_get_cwm(QUIC_TXFC *txfc)
+{
+    return txfc->cwm;
+}
+
+uint64_t ossl_quic_txfc_get_swm(QUIC_TXFC *txfc)
+{
+    return txfc->swm;
+}
+
+/*
+ * RX Flow Controller (RXFC)
+ * =========================
+ */
+
+int ossl_quic_rxfc_init(QUIC_RXFC *rxfc, QUIC_RXFC *conn_rxfc,
+                        uint64_t initial_window_size,
+                        uint64_t max_window_size,
+                        OSSL_TIME (*now)(void *now_arg),
+                        void *now_arg)
+{
+    if (conn_rxfc != NULL && conn_rxfc->parent != NULL)
+        return 0;
+
+    rxfc->swm               = 0;
+    rxfc->cwm               = initial_window_size;
+    rxfc->rwm               = 0;
+    rxfc->esrwm             = 0;
+    rxfc->hwm               = 0;
+    rxfc->cur_window_size   = initial_window_size;
+    rxfc->max_window_size   = max_window_size;
+    rxfc->parent            = conn_rxfc;
+    rxfc->error_code        = 0;
+    rxfc->has_cwm_changed   = 0;
+    rxfc->epoch_start       = ossl_time_zero();
+    rxfc->now               = now;
+    rxfc->now_arg           = now_arg;
+    rxfc->is_fin            = 0;
+    return 1;
+}
+
+QUIC_RXFC *ossl_quic_rxfc_get_parent(QUIC_RXFC *rxfc)
+{
+    return rxfc->parent;
+}
+
+static void rxfc_start_epoch(QUIC_RXFC *rxfc)
+{
+    rxfc->epoch_start   = rxfc->now(rxfc->now_arg);
+    rxfc->esrwm         = rxfc->rwm;
+}
+
+static int on_rx_controlled_bytes(QUIC_RXFC *rxfc, uint64_t num_bytes)
+{
+    int ok = 1;
+    uint64_t credit = rxfc->cwm - rxfc->swm;
+
+    if (num_bytes > credit) {
+        ok = 0;
+        num_bytes = credit;
+        rxfc->error_code = QUIC_ERR_FLOW_CONTROL_ERROR;
+    }
+
+    rxfc->swm += num_bytes;
+    return ok;
+}
+
+int ossl_quic_rxfc_on_rx_stream_frame(QUIC_RXFC *rxfc, uint64_t end, int is_fin)
+{
+    uint64_t delta;
+
+    if (rxfc->parent == NULL)
+        return 0;
+
+    if (rxfc->is_fin && ((is_fin && rxfc->hwm != end) || end > rxfc->hwm)) {
+        /* Stream size cannot change after the stream is finished */
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
```

Comment:
Heh, now I can delete all the fin handling in my receive buffers implementation. :grin: